Added ZippedRange, and ported some improvements from it to IndexedRange

geoffreygaren opened the pull request:
#### 15e2f508ca1a41912a0d13c510a4c5704883f2b9
<pre>
Added ZippedRange, and ported some improvements from it to IndexedRange
<a href="https://bugs.webkit.org/show_bug.cgi?id=283249">https://bugs.webkit.org/show_bug.cgi?id=283249</a>
<a href="https://rdar.apple.com/140054429">rdar://140054429</a>

Reviewed by Chris Dumez.

We use these types to assist bounds-safe iteration.

* Source/WTF/WTF.xcodeproj/project.pbxproj: New header.

* Source/WTF/wtf/CMakeLists.txt: New header.

* Source/WTF/wtf/IndexedRange.h:
(WTF::BoundsCheckedIterator::begin):
(WTF::BoundsCheckedIterator::end): Use a universal reference argument type and
std::forward so that we can handle const / non-const and other variations.
Previously we only supported const&amp;, but sometimes you want to assign while
you iterate.

(WTF::BoundsCheckedIterator::operator* const): Support non-copyable types and
avoid expensive copies by returning the universal reference type.

&apos;auto&apos; return values decay &quot;T&amp;&quot; to &quot;T&quot;, which often creates a copy. &apos;auto&amp;&amp;&apos;
does not decay, and returns &quot;T&amp;&quot; for &quot;T&amp;&quot;.

If you are as confused as I am, these sources may help you:

<a href="https://skebanga.github.io/structured-bindings/">https://skebanga.github.io/structured-bindings/</a>
<a href="https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369">https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369</a>
<a href="https://devblogs.microsoft.com/oldnewthing/20201014-00/?p=104367">https://devblogs.microsoft.com/oldnewthing/20201014-00/?p=104367</a>

(WTF::BoundsCheckedIterator::BoundsCheckedIterator):
(WTF::boundsCheckedBegin):
(WTF::boundsCheckedEnd): No need to have Collection as a part of our type;
it&apos;s only needed in our constructor.

(WTF::IndexedRangeIterator::operator*): Just use pair. Tuple is more fancy
than we need to be.

(WTF::IndexedRange::IndexedRange): Use the universal reference type.

(WTF::indexedRange): We need to use a helper function instead of instantiating
IndexedRange directly in order to trigger template argument deduction.

* Source/WTF/wtf/JSONValues.cpp: Initial use of ZippedRange.

* Source/WTF/wtf/ZippedRange.h: Added. Same as IndexedRange, except we keep
two iterators instead of an index and an iterator. There are fancier ways to
do this, but this seems to fit our needs.

Here&apos;s one fancier option:
<a href="https://github.com/alemuntoni/zip-views">https://github.com/alemuntoni/zip-views</a>

(WTF::ZippedRangeIterator::ZippedRangeIterator):
(WTF::ZippedRangeIterator::operator++):
(WTF::ZippedRangeIterator::operator*):
(WTF::ZippedRangeIterator::operator== const):
(WTF::ZippedRange::ZippedRange):
(WTF::ZippedRange::begin):
(WTF::ZippedRange::end):
(WTF::zippedRange):

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::Device::createRenderBundleEncoder):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):
* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.mm:
(-[_WKTouchEventGenerator init]):
(-[_WKTouchEventGenerator _updateTouchPoints:window:]):
(-[_WKTouchEventGenerator touchDownAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator liftUpAtPoints:touchCount:window:]): Updated to use
indexedRange() instead of IndexedRange(), since indexedRange() is now required
for template argument deduction.

Canonical link: <a href="https://commits.webkit.org/286698@main">https://commits.webkit.org/286698@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1afc9c141787df4d01a78bc6b61e7bfaf75c7596

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55863 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81366 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28104 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65005 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4156 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60195 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18285 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79895 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50148 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23446 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26428 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70011 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82807 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76105 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4204 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2785 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68477 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67731 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11713 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9794 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98359 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11888 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4151 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6963 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21515 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4174 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7605 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5932 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ZippedRange.h:
```diff
@@ -0,0 +1,96 @@
+template<typename IteratorA, typename IteratorB> class ZippedRangeIterator {
+public:
+    ZippedRangeIterator(IteratorA&& iteratorA, IteratorB&& iteratorB)
+        : m_iteratorA(WTFMove(iteratorA))
```
Shouldn't this be std::forward because that `IteratorA` is a template parameter? If the caller didn't pass an-r-value reference, moving the value out would be confusing / error-prone.

geoffreygaren replied:
Apologies if I have this wrong, because it's new to me, but here's my understanding:

In the context of `ZippedRangeIterator<IteratorA, IteratorB>::ZippedRangeIterator<IteratorA, IteratorB>()`, `IteratorA` and `IteratorB` are known types, not deduced types. Therefore, `IteratorA&&` and `IteratorB&&` mean rvalue-reference, and not universal reference. Therefore, the caller is required to pass us an rvalue-reference.

(In contrast, if we had a free function `template<typename IteratorA, typename IteratorB> makeZippedRangeIterator(IteratorA&& iteratorA, IteratorB&& iteratorB)`, in the context of that free function `IteratorA` and `IteratorB` would not be known types, and would be deduced types, and `IteratorA&&` and `IteratorB&&` would mean universal reference.)

cdumez commented on Source/WTF/wtf/ZippedRange.h:
```diff
@@ -0,0 +1,96 @@
+public:
+    ZippedRangeIterator(IteratorA&& iteratorA, IteratorB&& iteratorB)
+        : m_iteratorA(WTFMove(iteratorA))
+        , m_iteratorB(WTFMove(iteratorB))
```
ditto.

geoffreygaren committed:
Added ZippedRange, and ported some improvements from it to IndexedRange
https://bugs.webkit.org/show_bug.cgi?id=283249
rdar://140054429

Reviewed by Chris Dumez.

We use these types to assist bounds-safe iteration.

* Source/WTF/WTF.xcodeproj/project.pbxproj: New header.

* Source/WTF/wtf/CMakeLists.txt: New header.

* Source/WTF/wtf/IndexedRange.h:
(WTF::BoundsCheckedIterator::begin):
(WTF::BoundsCheckedIterator::end): Use a universal reference argument type and
std::forward so that we can handle const / non-const and other variations.
Previously we only supported const&, but sometimes you want to assign while
you iterate.

(WTF::BoundsCheckedIterator::operator* const): Support non-copyable types and
avoid expensive copies by returning the universal reference type.

'auto' return values decay "T&" to "T", which often creates a copy. 'auto&&'
does not decay, and returns "T&" for "T&".

If you are as confused as I am, these sources may help you:

https://skebanga.github.io/structured-bindings/
https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369
https://devblogs.microsoft.com/oldnewthing/20201014-00/?p=104367

(WTF::BoundsCheckedIterator::BoundsCheckedIterator):
(WTF::boundsCheckedBegin):
(WTF::boundsCheckedEnd): No need to have Collection as a part of our type;
it's only needed in our constructor.

(WTF::IndexedRangeIterator::operator*): Just use pair. Tuple is more fancy
than we need to be.

(WTF::IndexedRange::IndexedRange): Use the universal reference type.

(WTF::indexedRange): We need to use a helper function instead of instantiating
IndexedRange directly in order to trigger template argument deduction.

* Source/WTF/wtf/JSONValues.cpp: Initial use of ZippedRange.

* Source/WTF/wtf/ZippedRange.h: Added. Same as IndexedRange, except we keep
two iterators instead of an index and an iterator. There are fancier ways to
do this, but this seems to fit our needs.

Here's one fancier option:
https://github.com/alemuntoni/zip-views

(WTF::ZippedRangeIterator::ZippedRangeIterator):
(WTF::ZippedRangeIterator::operator++):
(WTF::ZippedRangeIterator::operator*):
(WTF::ZippedRangeIterator::operator== const):
(WTF::ZippedRange::ZippedRange):
(WTF::ZippedRange::begin):
(WTF::ZippedRange::end):
(WTF::zippedRange):

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::Device::createRenderBundleEncoder):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::m_maxDrawCount):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):
* Source/WebKit/UIProcess/API/Cocoa/_WKTouchEventGenerator.mm:
(-[_WKTouchEventGenerator init]):
(-[_WKTouchEventGenerator _updateTouchPoints:window:]):
(-[_WKTouchEventGenerator touchDownAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator liftUpAtPoints:touchCount:window:]): Updated to use
indexedRange() instead of IndexedRange(), since indexedRange() is now required
for template argument deduction.

Canonical link: https://commits.webkit.org/286698@main
